server/Controllers/GameControllers/snake.js

```diff
@@ -1,13 +1,126 @@
 class Snake {
+  /**
+   * @param  {string} playerId
+   * @param  {number} initialPositionX
+   * @param  {number} initialPositionY
+   * @param  {string} initialDirection
+   * @param  {string} snakeColor
+   * @param  {number} initialSnakeLength
+   * @param  {Function} eatFood Function to check that snake's next movement will be eating food.
+   * @param  {Function} hasCollidedWithRival Function to check that snake's next movement is colliding with Rival's body or not
+   * @param  {number} intervalNumber
+   */
   constructor(playerId, initialPositionX, initialPositionY,
-    initialDirection, snakeColor, initialSnakeLength) {
+    initialDirection, snakeColor, initialSnakeLength, eatFood, hasCollidedWithRival, intervalNumber) {
     this.ownerId = playerId;
     this.head = { x: initialPositionX, y: initialPositionY };
     this.direction = initialDirection;
     this.color = snakeColor;
     this.bodyLength = initialSnakeLength;
+    this.eatFood = eatFood;
+    this.hasCollidedWithRival = hasCollidedWithRival;
```

Comment:
Instead of declaring eatfood in parent and passing it here, you can declare the function here only. and call it from parent by passing food co-ordinates..

Reply:
Here's an example: 

`js

class Snake {
  construction() {}

  moveSnake() {
    // move snake
  }

  eatFood(foodLocation) {
    // check if headlocationmatches
  }

  hasCollidedWith(otherSnake) {
    // check if has collided with otherSnake
  }
}

class Game {
  constructor() {
    this.food = {}; // food location
    this.snake1 = {};
    this.snake2 = {};
  }

  moveSnakes() {
    snake1.moveSnake();
    snake1.eatFood(this.food);
    snake1.hasCollidedWith(this.snake2);
    snake2.moveSnake();
    snake2.eatFood(this.food);
    snake2.hasCollidedWith(this.snake1);
  }
}
`

Reply:
@niinpatel but the when the food has been eaten, we need to create it again. So if keeep that eatFood function in the game class, we will create food there. Before returning true or false for eatFood() method call in the snake class.

Reply:
we can create he food only if food has been eaten:
```js
 if (snake1.eatFood(this.food)) {
      snake1.grow();
      this.createFood();
 }

Reply:
I have changed the structure, We will give body coordinates of the rival to the snake. And it will check itself the collision with the rival. And In my opinion, we shouldn't call grow function outside of the class.